### PR TITLE
Add images to canvas nodes

### PR DESCRIPTION
**Pull request by mdroidian**
https://www.loom.com/share/223bbfa9ee164607acb8d408e31f22c3

I noticed while making this video that the copyAsPNG/SVG sometimes doesn't align the image/text correctly on the y axis.
Also, copy as SVG creates an extremely large string that will break roam sync if pasted back into the canvas.

### CODE REVIEW

File: src/components/CanvasSettings.tsx [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    () => getSettingValueFromTree({ tree, key: "key-image" }) === "true"
+  );
+  const [keyImageOption, setKeyImageOption] = useState(() =>
+    getSettingValueFromTree({ tree, key: "key-image-option" })
+  );
+  const smartblockTemplateNode = useMemo(
+    () => getSubTree({ parentUid: uid, tree, key: "smartblock-template" }),
+    [uid, tree]
+  );
+
```
Comment (dvargas92495) [inline src/components/CanvasSettings.tsx:28]: why a smartblock template over a query definition?

File: src/components/CanvasSettings.tsx [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+import React, { useEffect, useMemo, useRef, useState } from "react";
+import type { CustomField } from "roamjs-components/components/ConfigPanels/types";
+import getBasicTreeByParentUid from "roamjs-components/queries/getBasicTreeByParentUid";
+import getFirstChildUidByBlockUid from "roamjs-components/queries/getFirstChildUidByBlockUid";
+import getSettingValueFromTree from "roamjs-components/util/getSettingValueFromTree";
+import getSubTree from "roamjs-components/util/getSubTree";
+import setInputSetting from "roamjs-components/util/setInputSetting";
+import createBlock from "roamjs-components/writes/createBlock";
+
+const DiscourseNodeConfigPanel: CustomField["options"]["component"] = ({
```
Comment (dvargas92495) [inline src/components/CanvasSettings.tsx:11]: typo

Comment (dvargas92495) [reply src/components/CanvasSettings.tsx:11]: but actually it shouldn't be

Comment (dvargas92495) [reply src/components/CanvasSettings.tsx:11]: can define a global default

File: src/components/QueryDrawer.tsx [hunk 3]
```diff
@@ -25,6 +25,7 @@ import Export from "./Export";
 type Props = {
   blockUid: string;
   onloadArgs: OnloadArgs;
+  clearOnClick?: (tag: string) => Promise<void>;
```
Comment (dvargas92495) [inline src/components/QueryDrawer.tsx:28]: get rid of this if possible

Comment (dvargas92495) [reply src/components/QueryDrawer.tsx:28]: delete src/components/CytoscapePlayground.tsx lol

File: src/components/TldrawCanvas.tsx [hunk 4]
```diff
@@ -442,17 +445,188 @@ class DiscourseNodeUtil extends TLBoxUtil<DiscourseNodeShape> {
...
+        context.updateProps(shape.id, {
+          w,
+          h,
+        });
+
+        return;
+      }
+
+      const smartblocks = window.roamjs?.extension?.smartblocks;
+      console.log("smartblock", smartblocks);
```
Comment (dvargas92495) [inline src/components/TldrawCanvas.tsx:526]: .

File: src/index.ts [hunk 5]
```diff
@@ -511,20 +512,7 @@ svg.rs-svg-container {
...
-                `[:find ?uid :where [?b :block/uid ?uid] [or-join [?b] 
-                 [and [?b :block/string ?s] [[clojure.string/includes? ?s "{{query block:${queryRef}}}"]] ]
-                 ${getQueryPages(extensionAPI).map(
-                   (p) =>
-                     `[and [?b :node/title "${p.replace(/\*/, queryRef)}"]]`
-                 )}
-                  [and [?b :node/title "${queryRef}"]]
-            ]]`
-              )[0]
-              ?.toString() || "";
```
Comment (mdroidian) [inline src/index.ts:527]: used elsewhere so extracted to function

File: src/components/TldrawCanvas.tsx [hunk 6]
```diff
@@ -442,17 +475,51 @@ class DiscourseNodeUtil extends TLBoxUtil<DiscourseNodeShape> {
...
+    const setSizeAndImgProps = async ({
+      context,
+      text,
+      uid,
+    }: {
+      context: DiscourseNodeUtil;
+      text: string;
+      uid: string;
+    }) => {
+      if (!extensionAPI) return console.log("no extensionAPI");
```
Comment (dvargas92495) [inline src/components/TldrawCanvas.tsx:490]: .

File: src/utils/calcCanvasNodeSizeAndImg.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  } else {
+    imageUrl = getFirstImageByUid(uid);
+  }
+
+  const padding = Number(DEFAULT_STYLE_PROPS.padding.replace("px", ""));
+  const maxWidth = Number(MAX_WIDTH.replace("px", ""));
+  const effectiveWidth = maxWidth - 2 * padding;
+
+  try {
+    if (!imageUrl) throw new Error("No Image URL");
```
Comment (dvargas92495) [inline src/utils/calcCanvasNodeSizeAndImg.ts:91]: itd be nice to notify the user that an error happened (`renderToast`?)

Comment (dvargas92495) [reply src/utils/calcCanvasNodeSizeAndImg.ts:91]: if/else => intentional
try/catch => nonintentional

**Review by dvargas92495 — APPROVED** [review]

**Pull request merged**